src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1000,603 +907,315 @@ private ImmutableArray<Symbol> CreateMembers()
+                        continue;
+                    }
+                    else if (member is TupleElementFieldSymbol tupleField)
+                    {
+                        fields.Add(tupleField.UnderlyingField.OriginalDefinition);
+                    }
+                    else if (member is FieldSymbol field)
+                    {
+                        fields.Add(field.OriginalDefinition);
+                    }
                 }
-            }
 
-            return map;
+                Debug.Assert(fields.All(f => f is object));
+                return fields.ToImmutableAndFree();
+            }
         }
 
-        public TMember GetTupleMemberSymbolForUnderlyingMember<TMember>(TMember underlyingMemberOpt) where TMember : Symbol
+        private TypeSymbol MergeTupleNames(NamedTypeSymbol other, NamedTypeSymbol mergedType)
         {
-            if ((object)underlyingMemberOpt == null)
+            // Merge tuple element names, if any
+            ImmutableArray<string> names1 = TupleElementNames;
+            ImmutableArray<string> names2 = other.TupleElementNames;
+            ImmutableArray<string> mergedNames;
+            if (names1.IsDefault || names2.IsDefault)
             {
-                return null;
+                mergedNames = default;
             }
-
-            Symbol underlyingMemberDefinition = underlyingMemberOpt.OriginalDefinition;
-            if (TypeSymbol.Equals(underlyingMemberDefinition.ContainingType, _underlyingType.OriginalDefinition, TypeCompareKind.ConsiderEverything2))
+            else
             {
-                Symbol result;
-                if (UnderlyingDefinitionToMemberMap.TryGetValue(underlyingMemberDefinition, out result))
+                Debug.Assert(names1.Length == names2.Length);
+                mergedNames = names1.ZipAsArray(names2, (n1, n2) => string.CompareOrdinal(n1, n2) == 0 ? n1 : null)!;
+
+                if (mergedNames.All(n => n is null))
                 {
-                    return (TMember)result;
+                    mergedNames = default;
                 }
             }
 
-            return null;
-        }
-
-        public override ImmutableArray<Symbol> GetMembers(string name)
-        {
-            return GetMembers().WhereAsArray(m => m.Name == name);
+            bool namesUnchanged = mergedNames.IsDefault ? TupleElementNames.IsDefault : mergedNames.SequenceEqual(TupleElementNames);
+            return (namesUnchanged && this.Equals(mergedType, TypeCompareKind.ConsiderEverything))
+                ? this
+                : CreateTuple(mergedType, mergedNames, this.TupleErrorPositions, this.TupleElementLocations, this.Locations);
         }
 
-        public override ImmutableArray<NamedTypeSymbol> GetTypeMembers()
+        /// <summary>
+        /// The main purpose of this type is to store element names and also cache some information related to tuples.
+        /// </summary>
+        internal sealed class TupleUncommonData
```

Comment:
>TupleUncommonData [](start = 30, length = 17)

Should this be named `TupleData` instead? (It looks like every tuple type has an instance so it's not uncommon, at least not for tuple types.)

Reply:
Will rename `TupleExtraData`, because `TupleData` causes conflict with `TupleData` property.

---
In reply to: [356125483](https://github.com/dotnet/roslyn/pull/39370#discussion_r356125483) [](ancestors = 356125483)